### PR TITLE
nfc: nfc tag api

### PR DESCRIPTION
**Pull request by kapi-no**
High level API definition for NFC Tag emulators.

This change partially covers the scope of feature request described in #5703 issue.

### CODE REVIEW

**Comment by kapi-no** [comment]
CC: @carlescufi, @grochu

**Comment by nashif** [comment]
@sameo FYI

File: include/nfc/nfc.h [hunk 1]
```diff
@@ -0,0 +1,159 @@
...
+ *        completed.
+ *
+ * @param op type of operation
+ * @param err zero on success or (negative) error code otherwise.
+ */
+typedef void (*nfc_tag_op_completed_cb_t)(enum nfc_tag_op op, int err);
+
+/** @brief Set up NFC Tag.
+ *
+ *  Set up NFC Tag subsystem. Load NDEF message data.
```
Comment (sameo) [inline include/nfc/nfc.h]: Is that "loading" an NDEF into the NFC subsystem for future tag writing?

File: include/nfc/nfc.h [hunk 2]
```diff
@@ -0,0 +1,159 @@
...
+/** @brief Activate the NFC frontend.
+ *
+ *  After receiving a callback notification, events are posted to the
+ *  application callbacks.
+ *
+ *  @param[in] cb Callback to notify completion.
+ *
+ *  @return 0 on success or a negative error code on failure.
+ */
+int nfc_tag_enable(nfc_tag_op_completed_cb_t cb);
```
Comment (sameo) [inline include/nfc/nfc.h]: If you're planning to do more than NFC tags, shouldn't that be part of a higher level NFC API? i.e. why would the NFC front end management be linked to the NFC tag API?

File: include/nfc/nfc.h [hunk 3]
```diff
@@ -0,0 +1,159 @@
...
+ *  After receiving a callback notification, no more events will be posted to
+ *  the application callbacks.
+ *
+ *  @param[in] cb Callback to notify completion.
+ *
+ *  @return 0 on success or a negative error code on failure.
+ */
+int nfc_tag_disable(nfc_tag_op_completed_cb_t cb);
+
+/** @brief Set a new NDEF message in the enabled state.
```
Comment (sameo) [inline include/nfc/nfc.h]: What does `set` here refers to? How is that related to `nfc_tag_setup` ?

**Review by sameo — COMMENTED** [review]
I may be confused, but I'm missing a few points:

- As an application, how do I read an NFC tag? Is that handled through `nfc_tag_cb_register()` and then anytime a tag is detected, it gets automatically read and registered applications get notified?
- How do I write an NDEF into a tag? How do I write several different ones?
- Do we expect application to know what an NDEF is? i.e. is this API going to be extended to provide an NDEF parsing and building API?

**Comment by kapi-no** [comment]
Hi @sameo! Thank you for taking the time to review it.

As a general note, this API proposal is only intended for NFC tags (the NFC Forum Device in Listen Mode). It means that it cannot be used with NFC pollers (the NFC Forum Device in Poll Mode).

> - As an application, how do I read an NFC tag? Is that handled through nfc_tag_cb_register() and then anytime a tag is detected, it gets automatically read and registered applications get notified?

> - How do I write an NDEF into a tag? How do I write several different ones

As I pointed out, you cannot read an NFC tag using this API - reading and writing operations can only be performed by the NFC poller.

> - Do we expect application to know what an NDEF is? i.e. is this API going to be extended to provide an NDEF parsing and building API?

I imagine that we can have a separate module/library that would provide an API (i.e. ndef.h) for NDEF manipulation (building and parsing). 

After receiving NDEF data in the callback (i.e. ```ndef_data_written()```), the user could parse it at the application level using such a module. 

Similarly, the user could build the NDEF message with the NDEF module and then pass the generated NDEF data to ```nfc_tag_setup()``` to start NFC tag emulation or just update the current NDEF data when the emulation is already running (```nfc_tag_ndef_msg_set()``` function).

**Comment by kapi-no** [comment]
Hi @sameo! Thank you for taking the time to review it.

As a general note, this API proposal is only intended for NFC tags (the NFC Forum Device in Listen Mode). It means that it cannot be used with NFC pollers (the NFC Forum Device in Poll Mode).

> - As an application, how do I read an NFC tag? Is that handled through nfc_tag_cb_register() and then anytime a tag is detected, it gets automatically read and registered applications get notified?

> - How do I write an NDEF into a tag? How do I write several different ones

As I pointed out, you cannot read an NFC tag using this API - reading and writing operations can only be performed by the NFC poller.

> - Do we expect application to know what an NDEF is? i.e. is this API going to be extended to provide an NDEF parsing and building API?

I imagine that we can have a separate module/library that would provide an API (i.e. ndef.h) for NDEF manipulation (building and parsing). 

After receiving NDEF data in the callback (i.e. ```ndef_data_written()```), the user could parse it at the application level using such a module. 

Similarly, the user could build the NDEF message with the NDEF module and then pass the generated NDEF data to ```nfc_tag_setup()``` to start NFC tag emulation or just update the current NDEF data when the emulation is already running (```nfc_tag_ndef_msg_set()``` function).

**Comment by kapi-no** [comment]
Hi @sameo! Thank you for taking the time to review it.

As a general note, this API proposal is only intended for NFC tags (the NFC Forum Device in Listen Mode). It means that it cannot be used with NFC pollers (the NFC Forum Device in Poll Mode).

> - As an application, how do I read an NFC tag? Is that handled through nfc_tag_cb_register() and then anytime a tag is detected, it gets automatically read and registered applications get notified?

> - How do I write an NDEF into a tag? How do I write several different ones

As I pointed out, you cannot read an NFC tag using this API - reading and writing operations can only be performed by the NFC poller.

> - Do we expect application to know what an NDEF is? i.e. is this API going to be extended to provide an NDEF parsing and building API?

I imagine that we can have a separate module/library that would provide an API (i.e. ndef.h) for NDEF manipulation (building and parsing). 

After receiving NDEF data in the callback (i.e. ```ndef_data_written()```), the user could parse it at the application level using such a module. 

Similarly, the user could build the NDEF message with the NDEF module and then pass the generated NDEF data to ```nfc_tag_setup()``` to start NFC tag emulation or just update the current NDEF data when the emulation is already running (```nfc_tag_ndef_msg_set()``` function).

**Comment by kapi-no** [comment]
Hi @sameo! Thank you for taking the time to review it.

As a general note, this API proposal is only intended for NFC tags (the NFC Forum Device in Listen Mode). It means that it cannot be used with NFC pollers (the NFC Forum Device in Poll Mode).

> - As an application, how do I read an NFC tag? Is that handled through nfc_tag_cb_register() and then anytime a tag is detected, it gets automatically read and registered applications get notified?

> - How do I write an NDEF into a tag? How do I write several different ones

As I pointed out, you cannot read an NFC tag using this API - reading and writing operations can only be performed by the NFC poller.

> - Do we expect application to know what an NDEF is? i.e. is this API going to be extended to provide an NDEF parsing and building API?

I imagine that we can have a separate module/library that would provide an API (i.e. ndef.h) for NDEF manipulation (building and parsing). 

After receiving NDEF data in the callback (i.e. ```ndef_data_written()```), the user could parse it at the application level using such a module. 

Similarly, the user could build the NDEF message with the NDEF module and then pass the generated NDEF data to ```nfc_tag_setup()``` to start NFC tag emulation or just update the current NDEF data when the emulation is already running (```nfc_tag_ndef_msg_set()``` function).

**Comment by kapi-no** [comment]
Hi @sameo! Thank you for taking the time to review it.

As a general note, this API proposal is only intended for NFC tags (the NFC Forum Device in Listen Mode). It means that it cannot be used with NFC pollers (the NFC Forum Device in Poll Mode).

> - As an application, how do I read an NFC tag? Is that handled through nfc_tag_cb_register() and then anytime a tag is detected, it gets automatically read and registered applications get notified?

> - How do I write an NDEF into a tag? How do I write several different ones

As I pointed out, you cannot read an NFC tag using this API - reading and writing operations can only be performed by the NFC poller.

> - Do we expect application to know what an NDEF is? i.e. is this API going to be extended to provide an NDEF parsing and building API?

I imagine that we can have a separate module/library that would provide an API (i.e. ndef.h) for NDEF manipulation (building and parsing). 

After receiving NDEF data in the callback (i.e. ```ndef_data_written()```), the user could parse it at the application level using such a module. 

Similarly, the user could build the NDEF message with the NDEF module and then pass the generated NDEF data to ```nfc_tag_setup()``` to start NFC tag emulation or just update the current NDEF data when the emulation is already running (```nfc_tag_ndef_msg_set()``` function).

**Comment by kapi-no** [comment]
Hi @sameo! Thank you for taking the time to review it.

As a general note, this API proposal is only intended for NFC tags (the NFC Forum Device in Listen Mode). It means that it cannot be used with NFC pollers (the NFC Forum Device in Poll Mode).

> - As an application, how do I read an NFC tag? Is that handled through nfc_tag_cb_register() and then anytime a tag is detected, it gets automatically read and registered applications get notified?

> - How do I write an NDEF into a tag? How do I write several different ones

As I pointed out, you cannot read an NFC tag using this API - reading and writing operations can only be performed by the NFC poller.

> - Do we expect application to know what an NDEF is? i.e. is this API going to be extended to provide an NDEF parsing and building API?

I imagine that we can have a separate module/library that would provide an API (i.e. ndef.h) for NDEF manipulation (building and parsing). 

After receiving NDEF data in the callback (i.e. ```ndef_data_written()```), the user could parse it at the application level using such a module. 

Similarly, the user could build the NDEF message with the NDEF module and then pass the generated NDEF data to ```nfc_tag_setup()``` to start NFC tag emulation or just update the current NDEF data when the emulation is already running (```nfc_tag_ndef_msg_set()``` function).

**Comment by sameo** [comment]
> Hi @sameo! Thank you for taking the time to review it.

Sure, anytime.

> As a general note, this API proposal is only intended for NFC tags (the NFC Forum Device in Listen Mode). It means that it cannot be used with NFC pollers (the NFC Forum Device in Poll Mode).
> 

Ok, so this API is basically about emulating NFC tags with Zephyr, right?
I think this should be more clearly described from the documentation itself. I don't think Zephyr as an RTOS should be limited to emulating tags, and could very well extend to tag reading/writing and even p2p (it would certainly make sense for BT handover use cases), so I'd suggest to make it clear that this is only part of a potentially larger Zephyr NFC API and that it only covers tag emulation.

> > * As an application, how do I read an NFC tag? Is that handled through nfc_tag_cb_register() and then anytime a tag is detected, it gets automatically read and registered applications get notified?
> 
> > * How do I write an NDEF into a tag? How do I write several different ones
> 
> As I pointed out, you cannot read an NFC tag using this API - reading and writing operations can only be performed by the NFC poller.
>

I understand now. This API is basically loading an NDEF that the poller will read. And the poller could also write into our emulated NFC tag.
How do you specify which type of tags you want to emulate?

> > * Do we expect application to know what an NDEF is? i.e. is this API going to be extended to provide an NDEF parsing and building API?
> 
> I imagine that we can have a separate module/library that would provide an API (i.e. ndef.h) for NDEF manipulation (building and parsing).
> 
> After receiving NDEF data in the callback (i.e. `ndef_data_written()`), the user could parse it at the application level using such a module.
> 
> Similarly, the user could build the NDEF message with the NDEF module and then pass the generated NDEF data to `nfc_tag_setup()` to start NFC tag emulation or just update the current NDEF data when the emulation is already running (`nfc_tag_ndef_msg_set()` function).

That's the idea, yes. NDEF is a more complex spec than it looks and you don't want your API callers to all have their own implementation/interpretation of the spec.

**Comment by sameo** [comment]
> Hi @sameo! Thank you for taking the time to review it.

Sure, anytime.

> As a general note, this API proposal is only intended for NFC tags (the NFC Forum Device in Listen Mode). It means that it cannot be used with NFC pollers (the NFC Forum Device in Poll Mode).
> 

Ok, so this API is basically about emulating NFC tags with Zephyr, right?
I think this should be more clearly described from the documentation itself. I don't think Zephyr as an RTOS should be limited to emulating tags, and could very well extend to tag reading/writing and even p2p (it would certainly make sense for BT handover use cases), so I'd suggest to make it clear that this is only part of a potentially larger Zephyr NFC API and that it only covers tag emulation.

> > * As an application, how do I read an NFC tag? Is that handled through nfc_tag_cb_register() and then anytime a tag is detected, it gets automatically read and registered applications get notified?
> 
> > * How do I write an NDEF into a tag? How do I write several different ones
> 
> As I pointed out, you cannot read an NFC tag using this API - reading and writing operations can only be performed by the NFC poller.
>

I understand now. This API is basically loading an NDEF that the poller will read. And the poller could also write into our emulated NFC tag.
How do you specify which type of tags you want to emulate?

> > * Do we expect application to know what an NDEF is? i.e. is this API going to be extended to provide an NDEF parsing and building API?
> 
> I imagine that we can have a separate module/library that would provide an API (i.e. ndef.h) for NDEF manipulation (building and parsing).
> 
> After receiving NDEF data in the callback (i.e. `ndef_data_written()`), the user could parse it at the application level using such a module.
> 
> Similarly, the user could build the NDEF message with the NDEF module and then pass the generated NDEF data to `nfc_tag_setup()` to start NFC tag emulation or just update the current NDEF data when the emulation is already running (`nfc_tag_ndef_msg_set()` function).

That's the idea, yes. NDEF is a more complex spec than it looks and you don't want your API callers to all have their own implementation/interpretation of the spec.

**Comment by sameo** [comment]
> Hi @sameo! Thank you for taking the time to review it.

Sure, anytime.

> As a general note, this API proposal is only intended for NFC tags (the NFC Forum Device in Listen Mode). It means that it cannot be used with NFC pollers (the NFC Forum Device in Poll Mode).
> 

Ok, so this API is basically about emulating NFC tags with Zephyr, right?
I think this should be more clearly described from the documentation itself. I don't think Zephyr as an RTOS should be limited to emulating tags, and could very well extend to tag reading/writing and even p2p (it would certainly make sense for BT handover use cases), so I'd suggest to make it clear that this is only part of a potentially larger Zephyr NFC API and that it only covers tag emulation.

> > * As an application, how do I read an NFC tag? Is that handled through nfc_tag_cb_register() and then anytime a tag is detected, it gets automatically read and registered applications get notified?
> 
> > * How do I write an NDEF into a tag? How do I write several different ones
> 
> As I pointed out, you cannot read an NFC tag using this API - reading and writing operations can only be performed by the NFC poller.
>

I understand now. This API is basically loading an NDEF that the poller will read. And the poller could also write into our emulated NFC tag.
How do you specify which type of tags you want to emulate?

> > * Do we expect application to know what an NDEF is? i.e. is this API going to be extended to provide an NDEF parsing and building API?
> 
> I imagine that we can have a separate module/library that would provide an API (i.e. ndef.h) for NDEF manipulation (building and parsing).
> 
> After receiving NDEF data in the callback (i.e. `ndef_data_written()`), the user could parse it at the application level using such a module.
> 
> Similarly, the user could build the NDEF message with the NDEF module and then pass the generated NDEF data to `nfc_tag_setup()` to start NFC tag emulation or just update the current NDEF data when the emulation is already running (`nfc_tag_ndef_msg_set()` function).

That's the idea, yes. NDEF is a more complex spec than it looks and you don't want your API callers to all have their own implementation/interpretation of the spec.

**Comment by sameo** [comment]
> Hi @sameo! Thank you for taking the time to review it.

Sure, anytime.

> As a general note, this API proposal is only intended for NFC tags (the NFC Forum Device in Listen Mode). It means that it cannot be used with NFC pollers (the NFC Forum Device in Poll Mode).
> 

Ok, so this API is basically about emulating NFC tags with Zephyr, right?
I think this should be more clearly described from the documentation itself. I don't think Zephyr as an RTOS should be limited to emulating tags, and could very well extend to tag reading/writing and even p2p (it would certainly make sense for BT handover use cases), so I'd suggest to make it clear that this is only part of a potentially larger Zephyr NFC API and that it only covers tag emulation.

> > * As an application, how do I read an NFC tag? Is that handled through nfc_tag_cb_register() and then anytime a tag is detected, it gets automatically read and registered applications get notified?
> 
> > * How do I write an NDEF into a tag? How do I write several different ones
> 
> As I pointed out, you cannot read an NFC tag using this API - reading and writing operations can only be performed by the NFC poller.
>

I understand now. This API is basically loading an NDEF that the poller will read. And the poller could also write into our emulated NFC tag.
How do you specify which type of tags you want to emulate?

> > * Do we expect application to know what an NDEF is? i.e. is this API going to be extended to provide an NDEF parsing and building API?
> 
> I imagine that we can have a separate module/library that would provide an API (i.e. ndef.h) for NDEF manipulation (building and parsing).
> 
> After receiving NDEF data in the callback (i.e. `ndef_data_written()`), the user could parse it at the application level using such a module.
> 
> Similarly, the user could build the NDEF message with the NDEF module and then pass the generated NDEF data to `nfc_tag_setup()` to start NFC tag emulation or just update the current NDEF data when the emulation is already running (`nfc_tag_ndef_msg_set()` function).

That's the idea, yes. NDEF is a more complex spec than it looks and you don't want your API callers to all have their own implementation/interpretation of the spec.

**Comment by sameo** [comment]
> Hi @sameo! Thank you for taking the time to review it.

Sure, anytime.

> As a general note, this API proposal is only intended for NFC tags (the NFC Forum Device in Listen Mode). It means that it cannot be used with NFC pollers (the NFC Forum Device in Poll Mode).
> 

Ok, so this API is basically about emulating NFC tags with Zephyr, right?
I think this should be more clearly described from the documentation itself. I don't think Zephyr as an RTOS should be limited to emulating tags, and could very well extend to tag reading/writing and even p2p (it would certainly make sense for BT handover use cases), so I'd suggest to make it clear that this is only part of a potentially larger Zephyr NFC API and that it only covers tag emulation.

> > * As an application, how do I read an NFC tag? Is that handled through nfc_tag_cb_register() and then anytime a tag is detected, it gets automatically read and registered applications get notified?
> 
> > * How do I write an NDEF into a tag? How do I write several different ones
> 
> As I pointed out, you cannot read an NFC tag using this API - reading and writing operations can only be performed by the NFC poller.
>

I understand now. This API is basically loading an NDEF that the poller will read. And the poller could also write into our emulated NFC tag.
How do you specify which type of tags you want to emulate?

> > * Do we expect application to know what an NDEF is? i.e. is this API going to be extended to provide an NDEF parsing and building API?
> 
> I imagine that we can have a separate module/library that would provide an API (i.e. ndef.h) for NDEF manipulation (building and parsing).
> 
> After receiving NDEF data in the callback (i.e. `ndef_data_written()`), the user could parse it at the application level using such a module.
> 
> Similarly, the user could build the NDEF message with the NDEF module and then pass the generated NDEF data to `nfc_tag_setup()` to start NFC tag emulation or just update the current NDEF data when the emulation is already running (`nfc_tag_ndef_msg_set()` function).

That's the idea, yes. NDEF is a more complex spec than it looks and you don't want your API callers to all have their own implementation/interpretation of the spec.

**Comment by kapi-no** [comment]
Yes, that's right. 

> I think this should be more clearly described from the documentation itself. I don't think Zephyr as an RTOS should be limited to emulating tags, and could very well extend to tag reading/writing and even p2p (it would certainly make sense for BT handover use cases), so I'd suggest to make it clear that this is only part of a potentially larger Zephyr NFC API and that it only covers tag emulation.

I defined a Doxygen group called "NFC Tag API" to indicate that the API is intended for Tag emulation only, but docs improvements are always welcome. Any particular suggestions?

I imagine that additional API for NFC poller (or p2p) can be added later on to the existing NFC API. The current API can be viewed as its Bluetooth ```gatt.h``` counterpart without the GATT Client support.

> How do you specify which type of tags you want to emulate?

Hopefully, this API is generic enough to be compatible with any type of tag. When coming up with this proposal, I mostly analyzed Type 2 & 4 Tag platforms, because they are the most popular ones.

I imagine that user will choose the required tag platform via the Kconfig option, so future TxT (T2T, T4T, etc..) stack libraries need to implement this API.

File: include/nfc/nfc.h (hunk 1 above)
Comment (kapi-no) [reply include/nfc/nfc.h]: No, it is just initial NDEF data that will be used by the NFC tag after starting its emulation (with ```nfc_tag_enable()```)

File: include/nfc/nfc.h (hunk 2 above)
Comment (kapi-no) [reply include/nfc/nfc.h]: I am not planning to do more than NFC tags for now, but the description of the API function is a bit misleading/generic. I am going to change it to: "Activate the NFC tag emulation" (also for ```nfc_tag_disable()```).

File: include/nfc/nfc.h (hunk 3 above)
Comment (kapi-no) [reply include/nfc/nfc.h]: It can be used to update NDEF data dynamically when NFC tag emulation is already running.

**Comment by nashif** [comment]
stale, closing
